zipline/finance/performance/position_tracker.py

```diff
@@ -366,64 +365,73 @@ def _maybe_earn_dividend(self, dividend):
         else:
             return zp.dividend_payment()
 
-    def earn_dividends(self, dividend_frame):
+    def earn_dividends(self, dividends, stock_dividends):
         """
         Given a frame of dividends whose ex_dates are all the next trading day,
         calculate and store the cash and/or stock payments to be paid on each
         dividend's pay date.
         """
-        earned = dividend_frame.apply(self._maybe_earn_dividend, axis=1)\
-                               .dropna(how='all')
-        if len(earned) > 0:
+        for dividend in dividends:
             # Store the earned dividends so that they can be paid on the
             # dividends' pay_dates.
-            self._unpaid_dividends = pd.concat(
-                [self._unpaid_dividends, earned],
-            )
-
-    def _maybe_pay_dividend(self, dividend):
-        """
-        Take a historical dividend record, look up any stored record of
-        cash/stock we are owed for that dividend, and return a Series
-        with fields drawn from zipline.protocol.DIVIDEND_PAYMENT_FIELDS.
-        """
-        try:
-            unpaid_dividend = self._unpaid_dividends.loc[dividend['id']]
-            return unpaid_dividend
-        except KeyError:
-            return zp.dividend_payment()
-
-    def pay_dividends(self, dividend_frame):
+            div_owed = self.positions[dividend.sid].earn_dividend(dividend)
```

Comment:
The function that calls this gets the list of dividends based on the values in `self.positions`.

We could combine/inline `earn_dividends` into `check_upcoming_dividends` so that is an invariant.